src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/InlayHints/InlayHintService.cs

```diff
@@ -0,0 +1,126 @@
+            ProjectedRange: projectedRange,
+            ProjectedKind: RazorLanguageKind.CSharp
+        );
+
+        var inlayHints = await clientConnection.SendRequestAsync<DelegatedInlayHintParams, InlayHint[]?>(
+            CustomMessageNames.RazorInlayHintEndpoint,
+            delegatedRequest,
+            cancellationToken).ConfigureAwait(false);
+
+        if (inlayHints is null)
+        {
+            return null;
+        }
+
+        var csharpSourceText = codeDocument.GetCSharpSourceText();
+        using var _1 = ArrayBuilderPool<InlayHint>.GetPooledObject(out var inlayHintsBuilder);
+        foreach (var hint in inlayHints)
+        {
+            if (hint.Position.TryGetAbsoluteIndex(csharpSourceText, null, out var absoluteIndex) &&
+                _documentMappingService.TryMapToHostDocumentPosition(csharpDocument, absoluteIndex, out Position? hostDocumentPosition, out _))
+            {
+                if (hint.TextEdits is not null)
+                {
+                    hint.TextEdits = _documentMappingService.GetHostDocumentEdits(csharpDocument, hint.TextEdits);
+                }
+
+                hint.Data = new RazorInlayHintWrapper
+                {
+                    TextDocument = documentContext.Identifier,
+                    OriginalData = hint.Data,
+                    OriginalPosition = hint.Position
+                };
+                hint.Position = hostDocumentPosition;
+                inlayHintsBuilder.Add(hint);
+            }
+        }
+
+        return inlayHintsBuilder.ToArray();
+    }
+
+    public async Task<InlayHint?> ResolveInlayHintAsync(IClientConnection clientConnection, InlayHint inlayHint, CancellationToken cancellationToken)
+    {
+        var inlayHintWrapper = inlayHint.Data as RazorInlayHintWrapper;
+        if (inlayHintWrapper is null &&
+            inlayHint.Data is JObject dataObj)
+        {
+            inlayHintWrapper = dataObj.ToObject<RazorInlayHintWrapper>();
+        }
+
+        if (inlayHintWrapper is null)
+        {
+            return null;
+        }
+
+        var razorPosition = inlayHint.Position;
+        inlayHint.Position = inlayHintWrapper.OriginalPosition;
+        inlayHint.Data = inlayHintWrapper.OriginalData;
+
+        // For now we only support C# inlay hints. Once Web Tools adds support we'll need to inlayHint from both servers and combine
+        // the results, much like folding ranges.
+        var delegatedRequest = new DelegatedInlayHintResolveParams(
+            Identifier: inlayHintWrapper.TextDocument,
+            InlayHint: inlayHint,
+            ProjectedKind: RazorLanguageKind.CSharp
+        );
+
+        var resolvedHint = await clientConnection.SendRequestAsync<DelegatedInlayHintResolveParams, InlayHint?>(
+            CustomMessageNames.RazorInlayHintResolveEndpoint,
+            delegatedRequest,
+            cancellationToken).ConfigureAwait(false);
+
+        if (resolvedHint is null)
+        {
+            return null;
```

Comment:
>return null; [](http://example.com/codeflow?start=12&length=12)

Is it OK to fail to resolve or should we assert?

Reply:
Resolving just fills out the tooltip info, so I've no idea when it would fail, but it seems fine to me to do so. Potentially a delegated server could return null if there is no extra tooltip info (since its cheaper than serializing the whole InlayHint again)